mobile/asa-go/src/utils/PMTilesCache.ts

```diff
@@ -0,0 +1,169 @@
+  const base64Data = await blobToBase64(blob);
+  return base64Data;
+};
+
+const deserialize = (text: string): Blob => {
+  return base64ToBlob(text);
+};
+
+const toPMTiles = (file: ReadFileResult, filename: string) => {
+  const deserialized = deserialize(file.data as string);
+
+  // Initialize PMTiles with local blob
+  const pmtiles = new PMTiles(
+    new FileSource(new File([deserialized], filename))
+  );
+  return pmtiles;
+};
+
+const fetchAndStoreStaticPMTiles = (
+  filename: string,
+  fileSystem: FilesystemPlugin
+) => {
+  return async () => {
+    try {
+      const blob = await fetchStaticPMTiles(filename);
+      const serialized = await serialize(blob);
+
+      await fileSystem.writeFile({
+        path: filename,
+        data: serialized,
+        directory: Directory.Data,
+        encoding: Encoding.UTF8,
+      });
+
+      const file = await fileSystem.readFile({
+        path: filename,
+        directory: Directory.Data,
+        encoding: Encoding.UTF8,
+      });
+
+      return toPMTiles(file, filename);
+    } catch (error) {
+      console.error("Error storing PMTiles:", error);
+    }
+  };
+};
+
+const fetchAndStoreHFIPMTiles = (
+  for_date: DateTime,
+  run_type: RunType,
+  run_date: DateTime,
+  filename: string,
+  fileSystem: FilesystemPlugin
+) => {
+  return async () => {
+    try {
+      const blob = await fetchHFIPMTiles(for_date, run_type, run_date);
+      const serialized = await serialize(blob);
+
+      await fileSystem.writeFile({
+        path: filename,
+        data: serialized,
+        directory: Directory.Data,
+        encoding: Encoding.UTF8,
+      });
+
+      const file = await fileSystem.readFile({
+        path: filename,
+        directory: Directory.Data,
+        encoding: Encoding.UTF8,
+      });
+
+      return toPMTiles(file, filename);
+    } catch (error) {
+      console.error("Error storing PMTiles:", error);
+    }
+  };
+};
+
+export interface IPMTilesCache {
+  loadPMTiles: (
+    filename: string,
+    fetchAndStoreCallback?: () => Promise<PMTiles | undefined>
+  ) => Promise<PMTiles | undefined>;
+  loadHFIPMTiles: (
+    for_date: DateTime,
+    run_type: RunType,
+    run_date: DateTime,
+    filename: string
+  ) => Promise<PMTiles | undefined>;
+}
+
+export class PMTilesCache implements IPMTilesCache {
+  constructor(private readonly fileSystem: FilesystemPlugin) {}
+  public readonly loadPMTiles = async (
+    filename: string,
+    fetchAndStoreCallback?: () => Promise<PMTiles | undefined>
+  ) => {
+    const fetchAndStore =
+      fetchAndStoreCallback ??
+      fetchAndStoreStaticPMTiles(filename, this.fileSystem);
+    try {
+      const file = await this.fileSystem.readFile({
+        path: filename,
+        directory: Directory.Data,
+        encoding: Encoding.UTF8,
+      });
+
+      return toPMTiles(file, filename);
+    } catch (e) {
+      console.log("Error reading file, attempting to re-fetch", e);
+      const pmTiles = await fetchAndStore();
```

Comment:
Do you think this could get stuck in an endless loop? If so, we might consider limiting it to specific number of retries?

Reply:
Yeah good point, I'll adjust this.

Reply:
Added in https://github.com/bcgov/wps/pull/4238/commits/c7bccb809c17b3e4ed31a0a626289e13637b3d50
with tests adjusted in https://github.com/bcgov/wps/pull/4238/commits/384a7461b76bb12633b1c1e60c30b6faa084cc82 and https://github.com/bcgov/wps/pull/4238/commits/2a6524252f35f9c32593bde75552d7535cc2cdf0